### PR TITLE
GitHub: cache more things

### PR DESCRIPTION
**Pull request by garymm**
also remove unnecessary remote_download_minimal. We're not using a
remote cache so it's pointless.

Change-Id: I375b456a064ccbd57d6cdc94f9b548371d92ddaf

### CODE REVIEW

**Comment by garymm** [comment]
@oliverlee I'm a bit worried about this in the post-cache step:
`Warning: Path Validation Error: Path(s) specified in the action for caching do(es) not exist, hence no cache is being saved.`

I tried to reproduce locally but couldn't. Maybe it will be fixed once this runs on master and the cache is created.

**Comment by oliverlee** [comment]
Github actions cache cannot be updated once it is created (except for separate versions on feature branches).
https://github.com/actions/cache/blob/main/tips-and-workarounds.md#update-a-cache

This is how I handle repo cache in geometry:
https://github.com/oliverlee/geometry/commit/0ada9e3837c5762e974692bf25af5987abd1dcee
https://github.com/oliverlee/geometry/commit/c3c0a3408fee89a4f3f24175604a6b39df55c127
The jobs run in the `check` workflow do not save repo cache.

From looking at profiling data in CI logs, `ctx.download_and_extract` of the llvm toolchain (which occurs in almost all CI jobs) takes ~2 mins and dominates the load+analyze phases. By setting up `actions/cache/restore` (which takes ~ 1 min), the load+analyze phase is reduced to a few seconds and results in ~1 minute saved overall.

Here's a snippet from a recent build:

```
=== PHASE SUMMARY INFORMATION ===

Total launch phase time                              1.581 s   11.32%
Total init phase time                                1.769 s   12.67%
Total target pattern evaluation phase time           0.254 s    1.82%
Total interleaved loading, analysis and execution phase time   10.326 s   73.93%
Total finish phase time                              0.035 s    0.25%
---------------------------------------------------------------------
Total run time                                      13.966 s  100.00%
```
https://github.com/oliverlee/geometry/actions/runs/7514778513/job/20458026966

**Comment by oliverlee** [comment]
Also, the single saved repo cache in geometry is ~4GB. I assume it will be similar here if successful since the same hermetic toolchains are used. Since each job is creating a separate cache entry, each workflow run will probably hit the cache limit and result in cache evictions.

https://github.com/actions/cache#cache-limits

File: .github/workflows/ci.bazelrc [hunk 1]
```diff
@@ -1,11 +1,14 @@
...
+build --show_timestamps
 build --terminal_columns=120
-build --remote_download_minimal
+
+common --disk_cache=~/.cache/bazel_disk_cache
+common --repository_cache=~/.cache/bazel_repository_cache
+
+startup --host_jvm_args=-Xmx2500m # avoid blowing up the RAM usage.
+startup --install_base=~/.cache/bazel_install_base
+startup --output_base=~/.cache/bazel_output_base
```
Comment (oliverlee) [inline .github/workflows/ci.bazelrc:11]: https://github.com/bazelbuild/bazel/issues/12892#issuecomment-1246172168

**Comment by garymm** [comment]
Hmm OK I guess not sure if this is the best, but seems a bit simpler that what you did to me. I split it into "bazel disk cache" which every job writes and reads, and "bazel other" which only one job writes and everyoone else just restores.

File: .github/workflows/ci.bazelrc (hunk 1 above)
Comment (garymm) [reply .github/workflows/ci.bazelrc]: Thanks. Fixed.

File: .github/workflows/check.yml [hunk 2]
```diff
@@ -13,115 +13,163 @@ jobs:
...
+        # just taking advantage of the fact that this one is
+        # not a matrix and so only has one job.
+        uses: actions/cache@v3
+        with:
+          path: |
+            "~/.cache/bazel_install_base"
+            "~/.cache/bazel_output_base/external"
+            "~/.cache/bazel_repository_cache"
+            "~/.cache/bazelisk"
+          key: bazel-other
```
Comment (oliverlee) [inline .github/workflows/check.yml:80]: just fyi, you will only be able to write this cache once per branch.

**Review by oliverlee — APPROVED** [review]

File: .github/workflows/check.yml [hunk 3]
```diff
@@ -13,115 +13,163 @@ jobs:
...
-    - run: |
-        bazel \
-          --bazelrc=.github/workflows/ci.bazelrc \
-          test \
-          --config=${{ matrix.toolchain }} \
-          --features=${{ matrix.feature }} \
-          //...
+      - uses: actions/checkout@v3
+
+      - name: muont bazel disk cache
```
Comment (oliverlee) [inline .github/workflows/check.yml:21]:
```suggestion
      - name: mount bazel disk cache
```

**Comment by oliverlee** [comment]
Jobs show about 0 second duration for cache restore steps. Maybe you can try manually clearing the caches and then triggering the workflow twice?

File: .github/workflows/check.yml (hunk 2 above)
Comment (garymm) [reply .github/workflows/check.yml:80]: Thanks. I did not know that's how the cache action worked but I think that's fine for these.

**Pull request merged**